### PR TITLE
Empowered `strings::Join()` with predicates.

### PR DESCRIPTION
**Pull request by dkorolev**
@mzhurovich, PTAL. Thanks!

### CODE REVIEW

File: strings/split.h [hunk 1]
```diff
@@ -69,15 +69,30 @@ struct MatchImpl<std::string> {
 };
 
 template <size_t N>
-struct MatchImpl<const char[N]> {
+struct MatchImpl<const char[N]> {  // No `const char` because the calling semantics uses `rmconstref`.
```
Comment (mzhurovich) [inline strings/split.h]: Hm, this comment looks definitely not obvious to me :)
1) [No `const char`] next to `const char`
2) [calling semantics uses `rmconstref`], but I can't see any `rmconstref` below

File: strings/split.h [hunk 2]
```diff
@@ -69,15 +69,30 @@ struct MatchImpl<std::string> {
...
 
 template <typename T>
-inline bool Match(char a, T&& b) {
-  return MatchImpl<rmref<T>>::Match(a, b);
+static constexpr bool CanCallForChar(char) {
+  return false;
+}
+
+template <typename T>
+static constexpr auto CanCallForChar(int) -> decltype(std::declval<T>()(std::declval<char>()), bool()) {
```
Comment (mzhurovich) [inline strings/split.h]: SFINAE is addictive stuff :+1:

Comment (dkorolev) [inline strings/split.h]: It is. Got some? :+1:

File: strings/split.h (hunk 1 above)
Comment (dkorolev) [inline strings/split.h]: Oops. See the description of the very next commit.

File: strings/split.h (hunk 2 above)
Comment (mzhurovich) [inline strings/split.h]: This weed is the shit, dude!

**Comment by mzhurovich** [comment]
@dkorolev LGTM :+1:

**Comment by dkorolev** [comment]
Thanks @mzhurovich !

**Pull request merged**